### PR TITLE
Basic reference counting & pinning

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

A minimal working version of distributed reference counting - we pin the object at the local raylet when plasma objects are created and unpin them when they no longer have references at the owner or the owner dies.

The next step will be to cover serialized ObjectIDs, as this only considers those passed using the ray API.

TODO:
- [x] Fix failing `test_memory_limits.py` test
- [x] Comments in the code about how it works
- [x] Feature flag

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/protobuf/core_worker.proto
Comment (ericl) [inline src/ray/protobuf/core_worker.proto]: Keep this as the first arg for consistency?

File: src/ray/protobuf/core_worker.proto
Comment (ericl) [inline src/ray/protobuf/core_worker.proto]: Consider `WaitForObjectEviction` or similar.

File: python/ray/_raylet.pyx
Comment (zhijunfu) [reply python/ray/_raylet.pyx]: so if user specifies an object ID then it is not going to be pinned?

File: python/ray/worker.py
Comment (zhijunfu) [inline python/ray/worker.py]: I think it's easier to keep this code and just let the owner to pin the objects in plasma store, instead of going through raylet.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (zhijunfu) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Are we trying to do distributed reference counting here? Seems like the corresponding unnpin is not yet implemented, so this would end up creating objects that are not deletable? I suggest we postpone this until we have unpin ready, we can discuss more on this.

Also, It should be easier for the core worker to pin the return objects in this case, which is simpler to implement, and can avoid the issues mentioned above.

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: The timeout has to be zero, otherwise it's blocking raylet's event loop I think.

**Review by zhijunfu — CHANGES_REQUESTED** [review]
Is it possible to just let the owner to pin the objects instead in this case, so we don't need the extra RPCs between workers and raylet?

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: `XXX` is just me taking notes of things to address before merging

File: python/ray/tests/test_failure.py
Comment (edoakes) [reply python/ray/tests/test_failure.py]: Waits until the task finishes (ensures it doesn't hang in the sleep loop)

File: python/ray/_raylet.pyx
Comment (edoakes) [reply python/ray/_raylet.pyx]: Yes, added a comment. Note that this isn't a supported API.

File: python/ray/tests/test_memory_limits.py
Comment (edoakes) [reply python/ray/tests/test_memory_limits.py]: Ah sorry this shouldn't have been changed, snuck in. Memory limits still apply as they used to. These tests should still pass, debugging one of them.

File: python/ray/worker.py
Comment (edoakes) [reply python/ray/worker.py]: Ah yes that would be simpler, but it only works when the worker is on the same node as the objects that it owns. If a worker spawns a task that creates a remote plasma object, it is the owner of that object and needs to pin it but we don't want to have to transfer a copy of all owned plasma objects to the same node as the owner.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Unpin is implemented, it happens at the raylet when it gets a response to `WaitForObjectEviction`. This happens via a callback in the owner's reference counter when the object goes out of scope or if the worker crashes.

See above for why we're pinning at the raylet.

File: python/ray/_raylet.pyx
Comment (ericl) [reply python/ray/_raylet.pyx]: Can we raise an error here if ray_constants.direct_call_enabled()? I think the core worker will check fail in any case.

File: python/ray/tests/test_reference_counting.py
Comment (ericl) [inline python/ray/tests/test_reference_counting.py]: Don't need a ray.put here right?

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: AddDeleteCallback?

File: src/ray/core_worker/reference_count.cc
Comment (ericl) [inline src/ray/core_worker/reference_count.cc]: Seems like you need a list of callbacks here.

File: python/ray/tests/test_memory_limits.py
Comment (zhijunfu) [reply python/ray/tests/test_memory_limits.py]: thanks.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (zhijunfu) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: I see, thanks. I somehow misunderstood this part.

File: python/ray/worker.py
Comment (zhijunfu) [reply python/ray/worker.py]: Yes that's true. I meant that if worker is on the same then it's simpler and cheaper for the owner to pin it. In other cases we need raylet to do it. The drawback is we need two different solutions to handle different cases, but this is probably ok. What do you think?

**Comment by ericl** [comment]
@zhijunfu sorry for not having a more detailed design on this. Basically this is an early cut (unoptimized) for distributed reference counting. To resummarize, the strategy is as follows:

1. Objects are pinned by the local raylet when they are produced. This allows the lifetime of the object to exceed that of the worker, which is prone to dying, etc.
2. The raylet subscribes to the owner to know when it can unpin the object. This is currently done with individual long-running async RPCs.

I think we should aim to get a basic end-to-end implementation of ref counting working, and then see how we can optimize for performance after that. Since it only applies to large objects, it is likely some of the mechanisms used here are already fast enough.

File: src/ray/raylet/raylet_client.cc
Comment (ericl) [inline src/ray/raylet/raylet_client.cc]: Can we feature flag distributed ref counting? I think it would just need to make this function a no-op right?

File: python/ray/worker.py
Comment (edoakes) [reply python/ray/worker.py]: Ah I see your point now. We also discussed this but feel that it would be better to focus on simplicity and correctness for now by having a single code path and optimize later. Given that this code is only exercised for large objects, it's also unclear if these extra RPCs are a performance bottleneck.

File: python/ray/_raylet.pyx
Comment (edoakes) [reply python/ray/_raylet.pyx]: This is used in tests (it is a convenient way to synchronize when tasks finish)

File: src/ray/core_worker/reference_count.cc
Comment (edoakes) [reply src/ray/core_worker/reference_count.cc]: Not currently. Might in the future but will cross that bridge when we come to it

Comment (edoakes) [reply src/ray/core_worker/reference_count.cc]: Added a check to be sure we don't overwrite them. This will only ever be called once per object.

Comment (ericl) [reply src/ray/core_worker/reference_count.cc]: Won't you hit this if two workers wait for eviction on the same object?

**Review by ericl — APPROVED** [review]

**Comment by zhijunfu** [comment]
> @zhijunfu sorry for not having a more detailed design on this. Basically this is an early cut (unoptimized) for distributed reference counting.
> 
> I think we should aim to get a basic end-to-end implementation of ref counting working, and then see how we can optimize for performance after that. Since it only applies to large objects, it is likely some of the mechanisms used here are already fast enough.

@ericl Thanks for the update. Yes I think this makes sense. 

For distributed ref counting, How about we fleature-flag it until it's end-to-end ready?

**Comment by ericl** [comment]
@zhijunfu there is the object_pinning_enabled flag. It's on by default since we fall back to LRU, so it should be safe enough.

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: The `Get` here is a blocking call to plasma, and might can impacts on raylet since it's single threaded. Ideally we should use something that doesn't require a reply, or can wait the reply in a sync way. But we can revisit this later.

**Comment by ericl** [comment]
@edoakes can you run the microbenchmark before/after this change? There might be a noticeable difference.

File: python/ray/tests/test_reference_counting.py
Comment (zhijunfu) [inline python/ray/tests/test_reference_counting.py]: nit: this comment is no longer accurate.

File: python/ray/tests/test_reference_counting.py
Comment (zhijunfu) [inline python/ray/tests/test_reference_counting.py]: it might be more explicite to use `test_pending_task_return_object_pinning` instead?

File: src/ray/common/ray_config_def.h
Comment (zhijunfu) [inline src/ray/common/ray_config_def.h]: thanks!

File: src/ray/core_worker/core_worker.h
Comment (zhijunfu) [inline src/ray/core_worker/core_worker.h]: I'm trying to understand better on the flags. So for `Put` we always pin the object as it's owned by us. For `Seal` we still pin the objects for non-tests scenarios, where we own the object. It's only for tests that user can pass in an object id and in that case we won't pin the object. Is this the case? @edoakes

**Review by zhijunfu — APPROVED** [review]
Thanks, LGTM. I just added a few minor comments.

File: src/ray/core_worker/core_worker.h
Comment (edoakes) [reply src/ray/core_worker/core_worker.h]: I think one point of confusion is that `Put` is actually not used by Python at all and `Seal` is not used by Java at all - they each use the respective one for `ray.put()` calls. I only added this complex logic for `Seal` here because it's only used in Python tests but could add it for `Put` as well if you think that's better. Otherwise your summary is correct.

**Comment by edoakes** [comment]
@ericl @zhijunfu microbenchmark results.

The TL;DR is that it's substantially slower (~3x) for empty `ray.put()` calls. I'm not surprised that this is slow though it is worth than I expected. However, I don't think it's a big deal given that this should only be used for large objects. I also ran the benchmark with `10KiB` numpy arrays instead (the same as our threshold for spilling return objects), and the results are much better.

On master with zero-size objects:
```
single client get calls per second 49664.94 +- 473.27
single client put calls per second 10083.07 +- 26.14
single client put gigabytes per second 1.03 +- 0.67
multi client put calls per second 29614.41 +- 239.65
multi client put gigabytes per second 0.61 +- 0.1
single client tasks sync per second 2629.06 +- 12.78
single client tasks async per second 17038.58 +- 181.41
multi client tasks async per second 25093.0 +- 1227.91
1:1 actor calls sync per second 4594.93 +- 111.76
1:1 actor calls async per second 10459.96 +- 48.06
1:1 actor calls concurrent per second 12167.95 +- 49.66
1:n actor calls async per second 18976.5 +- 207.51
n:n actor calls async per second 31644.26 +- 1609.88
n:n actor calls with arg async per second 12484.84 +- 790.44
```

On this branch with zero-size objects:
```
single client get calls per second 21162.36 +- 729.11
single client put calls per second 9333.35 +- 176.0
single client put gigabytes per second 1.03 +- 0.66
multi client put calls per second 16062.86 +- 541.54
multi client put gigabytes per second 0.61 +- 0.09
single client tasks sync per second 2387.5 +- 21.19
single client tasks async per second 16468.02 +- 115.77
multi client tasks async per second 20068.04 +- 3585.0
1:1 actor calls sync per second 4159.72 +- 246.9
1:1 actor calls async per second 10054.61 +- 71.1
1:1 actor calls concurrent per second 11711.59 +- 68.09
1:n actor calls async per second 18340.54 +- 831.61
n:n actor calls async per second 29352.64 +- 1383.68
n:n actor calls with arg async per second 10142.2 +- 862.92
```

On master with `10KiB` numpy arrays:
```
single client get calls per second 26291.0 +- 1760.15
single client put calls per second 4256.05 +- 400.14
single client put gigabytes per second 0.95 +- 0.76
multi client put calls per second 10754.87 +- 6756.2
multi client put gigabytes per second 0.7 +- 0.08
single client tasks sync per second 2351.29 +- 82.59
single client tasks async per second 16505.24 +- 585.0
multi client tasks async per second 22332.28 +- 2264.2
1:1 actor calls sync per second 4558.48 +- 154.04
1:1 actor calls async per second 10408.01 +- 296.79
1:1 actor calls concurrent per second 10006.61 +- 95.26
1:n actor calls async per second 17238.29 +- 1025.11
n:n actor calls async per second 29930.91 +- 2465.03
n:n actor calls with arg async per second 11664.38 +- 923.27
```

On this branch with `10KiB` objects:
```
single client get calls per second 14004.17 +- 945.82
single client put calls per second 3929.97 +- 43.57
single client put gigabytes per second 0.94 +- 0.7
multi client put calls per second 7869.27 +- 4446.86
multi client put gigabytes per second 0.68 +- 0.06
single client tasks sync per second 2556.83 +- 63.46
single client tasks async per second 14944.14 +- 476.73
multi client tasks async per second 20339.06 +- 3797.85
1:1 actor calls sync per second 4019.7 +- 211.33
1:1 actor calls async per second 9439.17 +- 159.55
1:1 actor calls concurrent per second 9186.51 +- 173.44
1:n actor calls async per second 15831.67 +- 198.32
n:n actor calls async per second 26997.78 +- 1818.78
n:n actor calls with arg async per second 10500.02 +- 889.9
```

**Comment by ericl** [comment]
Hmm it actually looks like the get is slower not the put? That's odd.

File: src/ray/core_worker/core_worker.h
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.h]: I see, thanks. Yes that caused my confusion. I think it would be nice to also have it in java, so the features are more consistent for different user languages. If it's too late for this PR, it's also fine to do it in a later one.

**Comment by zhijunfu** [comment]
Thanks for the update. I think there could also be other cases where objects are put to plasma, which are not limited to large objects:
- when user calls ray.put
- when we forward the return objects of direct call to others by reference

Thus the actual frequency would depend on the exact user scenarios. 

> @ericl @zhijunfu microbenchmark results.
> 
> The TL;DR is that it's substantially slower (~3x) for empty `ray.put()` calls. I'm not surprised that this is slow though it is worth than I expected. However, I don't think it's a big deal given that this should only be used for large objects. I also ran the benchmark with `10KiB` numpy arrays instead (the same as our threshold for spilling return objects), and the results are much better.

**Comment by edoakes** [comment]
That is odd actually, not sure why that would be slower. I will try to look into it but will be slow because I'm away this week.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: I wonder if we'd really need the Delete/free interface with ref counting. Manual deletion doesn't interact well with the ref counting and in theory, you could just use `del x_id` in python to decrement the ref count and get the same effect.

File: src/ray/protobuf/core_worker.proto
Comment (stephanie-wang) [inline src/ray/protobuf/core_worker.proto]:
```suggestion
  // to this message indicates that the raylet should unpin the object.
```

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Delete?

File: src/ray/core_worker/reference_count.cc
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.cc]: Have you tried changing this if to an assertion? It seems like it should work now.

File: src/ray/core_worker/reference_count.cc
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.cc]: I think this TODO is no longer relevant since it's possible to add the first local reference before ownership info is added. Can you remove it and change it to a note that the ownership info must be added later in AddBorrowedObject?

File: src/ray/core_worker/reference_count.h
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.h]: Maybe it's obvious but could also add a note that this invalidates the iterator.

File: src/ray/core_worker/reference_count.h
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.h]:
```suggestion
  /// Get the owner ID and address of the given object.
```

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: You can set `it = worker_rpc_clients_.emplace(...).first` here.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Don't we need to reverse this line and the `Put`? Or else the raylet could try to get the object IDs before they are available in plasma.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Nit, but CoreWorker::ExecuteTask seems like a better place to put this code, since that is where objects are actually put into plasma (maybe after [this block](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L926)).

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: I would add a NOTE here that the caller must ensure that the objects are already in plasma before sending the request.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]:
```suggestion
  // Pin the requested objects until the owner notifies us that the objects can be unpinned.
```
```suggestion
  // Pin the requested objects.
```

**Review by stephanie-wang — APPROVED** [review]
Nice!

**Comment by stephanie-wang** [comment]
> Thanks for the update. I think there could also be other cases where objects are put to plasma, which are not limited to large objects:
> 
>     * when user calls ray.put

For small ray.puts, I think we could actually store the values in the in-memory store. I can't think of a reason why this wouldn't work.

File: src/ray/core_worker/core_worker.cc
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: Agreed, I think that would make sense. Given that we only expose this in the internal API I think we should consider deprecating it.

File: src/ray/core_worker/core_worker.cc
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: Good catch!

**Comment by zhijunfu** [comment]
I think this is doable, but if the object ids gets forwarded we still need to add it to plasma. Or are you suggesting that we inline these objects instead in the task spec? that might be doable too.  

> For small ray.puts, I think we could actually store the values in the in-memory store. I can't think of a reason why this wouldn't work.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Agree that it's probably better but started doing this and it's a little messy because we have to pipe some things around. Going to leave it as-is for now and we should clean this whole codepath up once the raylet execution path can be removed.

File: src/ray/core_worker/reference_count.cc
Comment (edoakes) [reply src/ray/core_worker/reference_count.cc]: This assertion fails in the case of randomly generated object ids, which are used in some tests.

Comment (edoakes) [reply src/ray/core_worker/reference_count.cc]: Ah actually it isn't failing from randomly generated object ids, but because we don't actually hold a python reference for large arguments that are passed implicitly by reference

**Comment by edoakes** [comment]
After further investigation, the slowdown on `ray.get()` is caused by no longer keeping the buffer reference (for our previous implementation of `weakref=True`). This causes the plasma client to make a `Release()` IPC in each loop in the benchmark now that it didn't used to.

Discussed with @pcmoritz and our preferred solution is to cache the buffer reference for `ObjectIDs` in the `ObjectID` after calling `ray.get` the first time. This should cover this microbenchmark case and also others (calling `ray.get` on the same ID many times in one process). Will leave it to a future PR, though.

**Comment by ericl** [comment]
Sounds fine to me.

**Pull request merged**
